lib/authenticate.js

```diff
@@ -15,119 +27,113 @@ const EXPIRES = `${THIS}:expires_at`;
-      return true;
-    }
+        storeToken(token$);
 
-    if (atom.config.get(REFRESH))
-    {
-      this._refresh();
-      return true;
+        return token$;
     }
 
-    return false;
-  }
+    return Observable.of(false);
+}
 
-  _isTokenExpired () {
-    let expires = atom.config.get(EXPIRES);
-
-    if (moment(expires).isSameOrBefore(moment()))
-      return true;
-
-    return false;
-  }
+function openBrowser() {
+    const URL = CLIENT.getAuthorizationUrl(STATE, REDIRECT_URI, [
+        MediumSDK.Scope.BASIC_PROFILE,
+        MediumSDK.Scope.PUBLISH_POST,
+    ]);
 
-  _refresh () {
-    let refreshToken = Encryption.decrypt(atom.config.get(REFRESH));
-    let _this = this;
-    return new Promise((resolve, reject) => {
-      _this.client.exchangeRefreshToken(refreshToken, (err, token) => {
-        if (err) reject(err);
-        resolve(_this._storeToken(token));
-      });
+    const { webContents } = (browser = new remote.BrowserWindow({
+        show: false,
+    }));
+
+    browser.loadURL(URL);
+    browser.show();
+
+    const selector = (event, url) => ({ event, url });
+
+    return Observable.fromEvent(webContents, 'will-navigate', selector)
+        .merge(
+            Observable.fromEvent(
+                webContents,
+                'did-get-redirect-request',
+                (event, oldURL, url) => ({ event, url })
+            )
+        )
+        .map(({ event, url }) => {
+            const rawCode = /code=([^&]*)/.exec(url) || null;
+            const code = rawCode && rawCode.length > 1 ? rawCode[1] : null;
+            const error = /\?error=(.+)$/.exec(url);
+
+            if (code || error) browser.destroy();
+
+            if (code) return code;
+            else if (error) throw new Error(error);
+        });
+}
+
+function storeToken(token$) {
+    token$.subscribe(({ access_token, refresh_token, expires_at }) => {
+        CLIENT.setAccessToken(access_token);
+        atom.config.set(ACCESS, Encryption.encrypt(access_token));
+        atom.config.set(REFRESH, Encryption.encrypt(refresh_token));
+        atom.config.set(EXPIRES, expires_at);
     });
-  }
-
-  _storeToken (token) {
-    this.client.setAccessToken(token.access_token);
-    atom.config.set(ACCESS, Encryption.encrypt(token.access_token));
-    atom.config.set(REFRESH, Encryption.encrypt(token.refresh_token));
-    atom.config.set(EXPIRES, token.expires_at);
-    return token;
-  }
-
-  _handleCallback (url) {
-    let rawCode = /code=([^&]*)/.exec(url) || null;
-    let code = (rawCode && rawCode.length > 1) ? rawCode[1] : null;
-    let error = /\?error=(.+)$/.exec(url);
-
-    if (code || error) this.browser.destroy();
-
-    if (code) this.emit('code', code);
-    else if (error) this.emit('error', error);
-  }
-};
+}
+
+export function authenticate() {
+    return hasToken()
+        .switchMap(
+            token =>
+                !token
+                    ? openBrowser().switchMap(code => {
```

Comment:
Is it worth moving this to a function that takes token. It's a bit hard to read in a ternary